File: node-server/src/services/keycloakService.ts
const jwt = require("jsonwebtoken");

class KeycloakService {
  private baseUrl: string;
  private realm: string;
  private clientId: string;
  private clientSecret: string;
  private tokenUrl: string;
  private userListUrl: string;

  private cachedToken: string | null = null;
  private tokenExpiry: number = 0;

  constructor() {
    this.baseUrl = process.env.KEYCLOAK_URL || "";
    this.realm = process.env.NODE_KEYCLOAK_REALM || "myrealm";
    this.clientId = process.env.NODE_CLIENT_ID || "nodeclient";
    this.clientSecret = process.env.NODE_CLIENT_SECRET || "";

    this.tokenUrl = `${this.baseUrl}/realms/${this.realm}/protocol/openid-connect/token`;
    this.userListUrl = `${this.baseUrl}/admin/realms/${this.realm}/users`;
  }

  private async getServiceAccountToken() {
    if (this.cachedToken && Date.now() < this.tokenExpiry) {
      return this.cachedToken;
    }

    const params = new URLSearchParams();
    params.append("grant_type", "client_credentials");
    params.append("client_id", this.clientId);
    params.append("client_secret", this.clientSecret);

    const res = await fetch(this.tokenUrl, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: params,
    });

    const data = await res.json();
    if (!data.access_token) throw new Error("Failed to get access token");

    const expiresIn = data.expires_in ? data.expires_in * 1000 : 60000;
    this.cachedToken = data.access_token;
    this.tokenExpiry = Date.now() + expiresIn - 5000;

    return this.cachedToken;
  }

  private async fetchWithAuth(url: string, options: RequestInit = {}) {
    const token = await this.getServiceAccountToken();

    const headers = {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
      ...(options.headers || {}),
    };

    return fetch(url, { ...options, headers });
  }

  async getUsers() {
    const res = await this.fetchWithAuth(this.userListUrl);
    if (!res.ok) throw new Error(`Failed to fetch users: ${await res.text()}`);
    return res.json();
  }

  async addRole(name: string, description: string) {
    const url = `${this.baseUrl}/admin/realms/${this.realm}/roles`;

    const res = await this.fetchWithAuth(url, {
      method: "POST",
      body: JSON.stringify({ name, description }),
    });

    if (res.status === 201) return { success: true };

    const text = await res.text();
    console.error("Failed to create role:", res.status, text);
    return { success: false, message: text };
  }

  async getRoles() {
    const res = await this.fetchWithAuth(`${this.baseUrl}/admin/realms/${this.realm}/roles`);
    return res.json();
  }

  async getUserRole(userId: string) {
    const res = await this.fetchWithAuth(`${this.userListUrl}/${userId}/role-mappings`);
    return res.json();
  }

  async getUserAvailableRole(userId: string) {
    const res = await this.fetchWithAuth(`${this.userListUrl}/${userId}/role-mappings/realm/available`);
    return res.json();
  }

  async addUserRole(userId: string, role: any) {
    const res = await this.fetchWithAuth(`${this.userListUrl}/${userId}/role-mappings/realm`, {
      method: "POST",
      body: JSON.stringify([role]),
    });
    return { success: res.status === 204 };
  }

  async removeUserRole(userId: string, role: any) {
    const res = await this.fetchWithAuth(`${this.userListUrl}/${userId}/role-mappings/realm`, {
      method: "DELETE",
      body: JSON.stringify([role]),
    });
    return { success: res.status === 204 };
  }

  decodeTokenData(token: string) {
    const rsa256key = `-----BEGIN PUBLIC KEY-----\n${process.env.RSA_256_KEY}\n-----END PUBLIC KEY-----\n`;
    return jwt.verify(token, rsa256key, { algorithms: ["RS256"] });
  }
}

export const keycloakService = new KeycloakService();

File: aspnetapp/KeycloakService.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.IdentityModel.Tokens;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Cryptography;

public class KeycloakService
{
    private readonly string _baseUrl;
    private readonly string _realm;
    private readonly string _clientId;
    private readonly string _clientSecret;

    private readonly string _tokenUrl;
    private readonly string _userListUrl;

    private string _cachedToken;
    private DateTime _tokenExpiry;

    private readonly HttpClient _httpClient;

    public KeycloakService()
    {
        _baseUrl = "http://localhost:8080";
        _realm =  "myrealm";
        _clientId = "nodeclient";
        _clientSecret = "v8Z9mTdKACG7rORIlWBMVOl2y9n9D2xC";

        _tokenUrl = $"{_baseUrl}/realms/{_realm}/protocol/openid-connect/token";
        _userListUrl = $"{_baseUrl}/admin/realms/{_realm}/users";

        _httpClient = new HttpClient();
    }

    private async Task<string> GetServiceAccountTokenAsync()
    {
        if (!string.IsNullOrEmpty(_cachedToken) && DateTime.UtcNow < _tokenExpiry)
        {
            return _cachedToken;
        }

        var content = new FormUrlEncodedContent(new[]
        {
            new KeyValuePair<string, string>("grant_type", "client_credentials"),
            new KeyValuePair<string, string>("client_id", _clientId),
            new KeyValuePair<string, string>("client_secret", _clientSecret)
        });

        var response = await _httpClient.PostAsync(_tokenUrl, content);
        response.EnsureSuccessStatusCode();

        var json = await response.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;

        if (!root.TryGetProperty("access_token", out var accessToken))
            throw new Exception("Failed to get access token");

        _cachedToken = accessToken.GetString();
        int expiresIn = root.TryGetProperty("expires_in", out var exp) ? exp.GetInt32() : 60;
        _tokenExpiry = DateTime.UtcNow.AddSeconds(expiresIn - 5); // 5 sec buffer

        return _cachedToken;
    }

    private async Task<HttpResponseMessage> FetchWithAuthAsync(string url, HttpMethod method, object body = null)
    {
        var token = await GetServiceAccountTokenAsync();
        var request = new HttpRequestMessage(method, url);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        if (body != null)
        {
            var json = JsonSerializer.Serialize(body);
            request.Content = new StringContent(json, Encoding.UTF8, "application/json");
        }

        return await _httpClient.SendAsync(request);
    }

    public async Task<JsonElement> GetUsersAsync()
    {
        var res = await FetchWithAuthAsync(_userListUrl, HttpMethod.Get);
        res.EnsureSuccessStatusCode();
        var json = await res.Content.ReadAsStringAsync();
        return JsonDocument.Parse(json).RootElement;
    }

    public async Task<bool> AddRoleAsync(string name, string description)
    {
        var url = $"{_baseUrl}/admin/realms/{_realm}/roles";
        var res = await FetchWithAuthAsync(url, HttpMethod.Post, new { name, description });

        return res.StatusCode == System.Net.HttpStatusCode.Created;
    }

    public async Task<JsonElement> GetRolesAsync()
    {
        var res = await FetchWithAuthAsync($"{_baseUrl}/admin/realms/{_realm}/roles", HttpMethod.Get);
        res.EnsureSuccessStatusCode();
        var json = await res.Content.ReadAsStringAsync();
        return JsonDocument.Parse(json).RootElement;
    }

    public async Task<JsonElement> GetUserRoleAsync(string userId)
    {
        var res = await FetchWithAuthAsync($"{_userListUrl}/{userId}/role-mappings", HttpMethod.Get);
        res.EnsureSuccessStatusCode();
        var json = await res.Content.ReadAsStringAsync();
        return JsonDocument.Parse(json).RootElement;
    }

    public async Task<JsonElement> GetUserAvailableRoleAsync(string userId)
    {
        var res = await FetchWithAuthAsync($"{_userListUrl}/{userId}/role-mappings/realm/available", HttpMethod.Get);
        res.EnsureSuccessStatusCode();
        var json = await res.Content.ReadAsStringAsync();
        return JsonDocument.Parse(json).RootElement;
    }

    public async Task<bool> AddUserRoleAsync(string userId, object role)
    {
        var res = await FetchWithAuthAsync($"{_userListUrl}/{userId}/role-mappings/realm", HttpMethod.Post, new[] { role });
        return res.StatusCode == System.Net.HttpStatusCode.NoContent;
    }

    public async Task<bool> RemoveUserRoleAsync(string userId, object role)
    {
        var res = await FetchWithAuthAsync($"{_userListUrl}/{userId}/role-mappings/realm", HttpMethod.Delete, new[] { role });
        return res.StatusCode == System.Net.HttpStatusCode.NoContent;
    }

    public JwtSecurityToken DecodeTokenData(string token)
    {
        var publicKeyPem = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvjuH9aBjjt4cmlhYZiARmARKU3GTulNcWF+BsWzE/rQOpltV3SlI70i56dQ9StJGyNTKFT89f64MyBz7lvHunj2fYG/WMCPkKXQKyvM7pr3bSVrU9xTRJh8czmGm9bY+12lyFvTn2SnfFAeAp9/hXBFw8IWemZvbEoF6lI+FYaN1MleIBY3IYZ0bMQOaBgFjaqRB45S8Hc0NK3lhEW2MqX04bZkD/7LRiBywZYENwkQ1eGt8i+fzeU9UxvX3nZU2HEnCGDjRuoUT24GJflK5BRWpiIwCGnPPZsxLgfM29DkRC3lIh319XUTsw01dZQgvhbVHjlktKrJBrcUWqnvgRwIDAQAB";
        var rsa = RSA.Create();
        rsa.ImportFromPem(publicKeyPem);

        var validationParameters = new TokenValidationParameters
        {
            ValidateIssuer = false,
            ValidateAudience = false,
            RequireSignedTokens = true,
            IssuerSigningKey = new RsaSecurityKey(rsa),
            ValidateLifetime = true
        };

        var handler = new JwtSecurityTokenHandler();
        handler.ValidateToken(token, validationParameters, out var validatedToken);
        return (JwtSecurityToken)validatedToken;
    }
}
